lib/Generate.js

```diff
@@ -152,9 +138,51 @@ function generateElmJson(
       return path.relative(generatedCodeDir, absolutePath);
     });
 
+  // Everything is cheap to calculate except dependencies. Given the same input
+  // we should produce the same output. This function has many inputs but
+  // basically only one output – an elm.json. So the easiest way to compare
+  // inputs is to generate as much of elm.json as possible – all the cheap parts
+  // – and then just add the remaining inputs: dependencies from the project elm.json.
+  // This way we can compare this “template” between runs and avoid unneeded
+  // expensive work.
+  let testElmJsonTemplate = {
+    type: 'application',
+    'source-directories': allSourceDirs,
+    'elm-version': '0.19.1',
+    // All dependencies will be replaced with resolved ones below.
+    dependencies: projectElmJson.dependencies,
+    'test-dependencies': projectElmJson['test-dependencies'],
+  };
+
+  const generatedPath = path.join(generatedCodeDir, 'elm.json');
+  const templatePath = path.join(generatedCodeDir, 'elm.template.json');
+  const templateElmJsonString = JSON.stringify(testElmJsonTemplate, null, 4);
+
+  try {
+    if (
+      // If the elm.json we are creating already exists…
+      fs.existsSync(generatedPath) &&
+      // …and its template hasn’t changed…
+      fs.readFileSync(templatePath, 'utf8') === templateElmJsonString
+    ) {
+      // …then just return early and go with the generated elm.json we already have.
+      // This avoids calculating exact dependencies again and saves a second or so.
+      return [generatedCodeDir, generatedSrc, sourceDirs];
+    }
+  } catch (_error) {
+    // No template file since before, moving on.
+  }
+
+  let testElmJson = Object.assign({}, testElmJsonTemplate, {
+    dependencies: Solve.get_dependencies(elmJsonPath, projectElmJson),
```

Comment:
Does `Solve.get_dependencies` not only take a single argument? https://github.com/rtfeldman/node-test-runner/blob/master/lib/solve.js#L3

Reply:
Correct, bad extraction from #442 (it had two arguments in an experiment over there)